modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_remote_rho_loss_downsampling.py

```diff
@@ -0,0 +1,109 @@
+    sampler = RemoteRHOLossDownsampling(**dummy_init_params)
+    assert sampler.per_sample_loss_fct == dummy_init_params["per_sample_loss"]
+    assert not sampler.requires_grad
+    mock_abstract_sampler_init__.assert_called_once_with(
+        dummy_init_params["pipeline_id"],
+        dummy_init_params["trigger_id"],
+        dummy_init_params["batch_size"],
+        dummy_init_params["params_from_selector"],
+        dummy_init_params["modyn_config"],
+        dummy_init_params["device"],
+    )
+
+
+@patch.object(IrreducibleLossProducer, "_load_il_model", return_value=dummy_model())
+def test_inform_samples(mock__load_il_model, dummy_init_params):
+    batch_size = 3
+
+    def fake_per_sample_loss(forward_output, target):
+        return torch.tensor(range(batch_size))
+
+    mock_per_sample_loss = Mock(wraps=fake_per_sample_loss)
+    dummy_init_params["per_sample_loss"] = mock_per_sample_loss
+    dummy_init_params["batch_size"] = batch_size
+    sampler = RemoteRHOLossDownsampling(**dummy_init_params)
+    sample_ids = list(range(batch_size))
+    forward_input = torch.randn(batch_size, 5)
+    forward_output = torch.randn(batch_size, 5)
+    target = torch.randint(5, (batch_size,))
+    embedding = None
+    fake_irreducible_loss = torch.tensor(range(batch_size, 0, -1))
+    with patch.object(
+        IrreducibleLossProducer, "get_irreducible_loss", return_value=fake_irreducible_loss
+    ) as mock_get_il:
+        sampler.init_downsampler()
+        sampler.inform_samples(sample_ids, forward_input, forward_output, target, embedding)
+        assert sampler.index_sampleid_map == sample_ids
+        assert sampler.rho_loss.shape == torch.Size([batch_size])
+        assert sampler.number_of_points_seen == 3
+        expected_rho_loss = torch.tensor([0, 1, 2]) - torch.tensor([3, 2, 1])
+        assert torch.allclose(sampler.rho_loss, expected_rho_loss)
+
+        mock_get_il.assert_called_once_with(sample_ids, ANY, ANY)
+        actual_forward_input = mock_get_il.call_args[0][1]
+        actual_target = mock_get_il.call_args[0][2]
+        assert torch.allclose(actual_forward_input, forward_input)
+        assert torch.allclose(actual_target, target)
+
+        mock_per_sample_loss.assert_called_once()
+        assert torch.allclose(mock_per_sample_loss.call_args[0][0], forward_output)
+        assert torch.allclose(mock_per_sample_loss.call_args[0][1], target)
+
+
+@patch.object(IrreducibleLossProducer, "_load_il_model", return_value=dummy_model())
+def test_select_points(mock__load_il_model, dummy_init_params):
+    dummy_init_params["batch_size"] = 5
+    dummy_init_params["params_from_selector"]["downsampling_ratio"] = 60
+    sampler = RemoteRHOLossDownsampling(**dummy_init_params)
+    sampler.rho_loss = torch.tensor([32, 7, 5, 8, 3])
+    sampler.number_of_points_seen = 5
+    sampler.index_sampleid_map = [2, 4, 1, 3, 5]
+    selected_ids, weights = sampler.select_points()
+    assert selected_ids == [2, 4, 3]
```

Comment:
should it not be [2,3,4] if it is sorted by rho loss? anyways i guess we shuffle, but i was just a little confused

Reply:
The current implementation fetches the top k but does not return the subset in the order sorted by rho loss, but in the oder consistent to the input order, i.e. here we have `[2, 4, 1, 3, 5]` as the input sample ids, so we keep the relative order. I think it's natural not to shuffle within the batch here? (As the only function is to select a subset).

But yes, it does not matter much I agree with you.